src/idl_gen_csharp.cpp

```diff
@@ -1745,93 +1763,107 @@ class CSharpGenerator : public BaseGenerator {
       auto is_array = IsArray(field.value.type);
       const auto &field_type =
           is_array ? field.value.type.VectorType() : field.value.type;
-      if (!IsStruct(field_type)) {
-        auto tmp_name_vec = name_vec;
-        tmp_name_vec.push_back(field.name);
-        auto tmp_array_length_vec = array_length_vec;
-        if (is_array) {
-          tmp_array_length_vec.push_back(field_type.fixed_length);
+      auto tmp_fields = fields;
+      FieldInfo tmp_field = {
+        field.name,
+        field_type.fixed_length,
+      };
+      tmp_fields.push_back(tmp_field);
+      if (field_type.struct_def != nullptr) {
+        GenStructPackDecl_ObjectAPI(*field_type.struct_def, code_ptr,
+                                    tmp_fields);
+      } else {
+        auto contains_array = false;
+        for (size_t i = 0; i < tmp_fields.size(); ++i) {
+          if (tmp_fields[i].array_length > 0) {
+            contains_array = true;
+            break;
+          }
         }
         std::string name;
-        for (size_t tmp_name_index = 0; tmp_name_index < tmp_name_vec.size();
-             ++tmp_name_index) {
-          name += "_" + tmp_name_vec[tmp_name_index];
+        for (size_t i = 0; i < tmp_fields.size(); ++i) {
+          name += "_" + tmp_fields[i].name;
         }
-        code += "    var " + name + " = new " + GenTypeBasic(field_type) + "[";
-        code += NumToString(tmp_array_length_vec[0]);
-        for (size_t i = 1; i < tmp_array_length_vec.size(); ++i) {
-          auto array_length = tmp_array_length_vec[i];
-          code += "," + NumToString(array_length);
-        }
-        code += "];\n";
-        code += "    ";
-        // initialize array
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          auto array_length = tmp_array_length_vec[i];
-          auto idx = "idx" + NumToString(i);
-          code += "for (var " + idx + " = 0; " + idx + " < " +
-                  NumToString(array_length) + "; ++" + idx + ") {";
-        }
-        code += name + "[idx0";
-        for (size_t i = 1; i < tmp_array_length_vec.size(); ++i) {
-          auto idx = "idx" + NumToString(i);
-          code += "," + idx;
-        }
-        code += "] = _o";
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          auto idx = "idx" + NumToString(i);
-          code += "." + MakeCamel(tmp_name_vec[i]) + "[" + idx + "]";
-        }
-        if (!is_array) { code += "." + MakeCamel(field.name); }
-        code += ";";
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          code += "}";
+        code += "    var " + name + " = ";
+        if (contains_array) {
+          code += "new " + GenTypeBasic(field_type) + "[";
+          auto first_array = true;
+          for (size_t i = 0; i < tmp_fields.size(); ++i) {
+            auto array_length = tmp_fields[i].array_length;
+            if (array_length <= 0) continue;
+            if (first_array) {
```

Comment:
You can replace this if-then and the variable with just `if (!i) code += ",";`

Reply:
This doesn't actually work unless we find out a way to not check for `<= 0`.